### PR TITLE
#36217: uint8 support for relational ops.

### PR DESCRIPTION
**Pull request by abdulla-TT**
### Ticket
#36217 

### Problem Description

- uint8 support is not available for relational ops.

### What's changed

- Added uint8 support for relational ops such as [LT, GT, LE, GE, NE, EQ].

Screenshot 2026-05-22 at 09 54 27

### Note for reviewrs
- [#44988](https://github.com/tenstorrent/tt-metal/issues/44988) A bug exists where typecasting uint8 to uint16 fails on tt-sim (Blackhole architecture) for multi-core tensors with shape [1, 3, 320, 320]. The failure is reproducible 100% of the time by running test_run_eltwise_typecast_op in tests/ttnn/unit_tests/operations/eltwise/test_eltwise_typecast.py. So added skip for BH tt-sim.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:abdulla/uint8_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:abdulla/uint8_support)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 1]
```diff
@@ -41,6 +41,75 @@
...
+        const std::optional<const DataType>& output_dtype,                                           \
+        const std::optional<MemoryConfig>& memory_config,                                            \
+        const std::optional<Tensor>& output,                                                         \
+        ttsl::Span<const operations::unary::EltwiseUnaryWithParam> post_activations,                 \
+        ttsl::Span<const operations::unary::EltwiseUnaryWithParam> lhs_activations,                  \
+        ttsl::Span<const operations::unary::EltwiseUnaryWithParam> rhs_activations,                  \
+        const std::optional<CoreRangeSet>& sub_core_grids,                                           \
+        const std::optional<tt::tt_metal::SubDeviceId>& sub_device_id) {                             \
+        const auto a = lhs.dtype() == DataType::UINT8 ? ttnn::typecast(lhs, DataType::UINT16) : lhs; \
+        const auto b = rhs.dtype() == DataType::UINT8 ? ttnn::typecast(rhs, DataType::UINT16) : rhs; \
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: **Nit: shallow Tensor copy on non-UINT8 path**

```cpp
const auto a = lhs.dtype() == DataType::UINT8 ? ttnn::typecast(lhs, DataType::UINT16) : lhs;
```

When `lhs` is not UINT8, `const auto a` triggers a Tensor copy-construction (the ternary returns by value). The old code passed `const Tensor& lhs` straight through. Tensor copies are cheap (shared storage), but you could avoid it with a helper or by using `const Tensor& a = lhs.dtype() == DataType::UINT8 ? ... : lhs;` — though you would need a named temporary for the typecast result to keep the reference valid. Not a blocker, just something to be aware of if this pattern spreads to hot paths.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 2]
```diff
@@ -41,6 +41,75 @@
...
+        return ttnn::detail::invoke_binary_ng(                                                       \
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: This line casts `rhs` independently from `lhs`, which is correct — a user could pass one UINT8 tensor and one UINT16 tensor, and both end up as UINT16. Worth confirming this mixed-dtype case (`UINT8 x UINT16`) is tested or intentionally unsupported. The current tests only exercise `UINT8 x UINT8`.

File: ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp [hunk 3]
```diff
@@ -749,6 +749,27 @@ tt::tt_metal::ProgramDescriptor BinaryNgDeviceOperation::ProgramFactory::create_
         use_llk_bcast = false;
     }
 
+    // Integer relational ops (NE/EQ/GT/LT/GE/LE) on UInt16 take the FPU SUB + NEZ
+    // postprocess path with DEST configured for Fp16_b accumulation (fp32_dest_acc_en
+    // is false for UInt16).  Under SCALAR broadcast the B2D datacopy unpacker writes
+    // a single u16 lane into all DEST positions; the resulting Fp16_b-tagged DEST is
+    // then read back by the NEZ postprocess, which interprets the integer bit pattern
+    // through the format-conversion path and corrupts the comparison result (#36217).
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp]: The comment says all six relational ops take "the FPU SUB + NEZ postprocess path", but only EQ/NE use SUB+{EQZ,NEZ} as postprocess — LT/GT/LE/GE go through the SFPU binary op path. The condition itself correctly handles both branches (via `postprocess.has_value()` for EQ/NE and the explicit `SfpuBinaryOp` checks for the others), so this is just a comment accuracy issue.

Suggested wording: "Integer relational ops on UInt16 use either the SUB + postprocess path (EQ/NE) or direct SFPU comparison (LT/GT/LE/GE). Under SCALAR broadcast, the B2D datacopy …"

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_uint8.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        (torch.Size([1, 2, 32, 64, 125]), torch.Size([1, 2, 32, 1, 1])),
+        (torch.Size([]), torch.Size([])),
+        (torch.Size([5]), torch.Size([1])),
+    ],
+)
+@pytest.mark.parametrize(
+    "low_a, high_a, low_b, high_b",
+    [
+        (0, 255, 0, 255),
+    ],
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_uint8.py]: The `low_a, high_a, low_b, high_b` parametrize has a single entry `(0, 255, 0, 255)`. Since there is only one value tuple, this just adds noise to the test ID without actually varying anything. Consider inlining `low=0, high=255` as constants in the test body, or adding more ranges (e.g., a narrow `(100, 200, 50, 150)`) to exercise partial-range comparisons.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_uint8.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    [ttnn.lt, ttnn.gt, ttnn.le, ttnn.ge, ttnn.ne, ttnn.eq],
+)
+def test_binary_relational_uint8_tensor_scalar(shape, scalar, ttnn_op, device):
+    # TODO: Remove this when typecasting uint8 to uint16 support is added to the BH simulator
+    # issue: https://github.com/tenstorrent/tt-metal/issues/44988
+    if is_blackhole() and os.environ.get("TT_METAL_SIMULATOR"):
+        pytest.skip("Skipping on BH tt-sim: UINT8->UINT16 typecast not supported")
+    num_elements = int(torch.prod(torch.tensor(shape)).item())
+    torch_input = torch.arange(num_elements, dtype=torch.int32).remainder(256).reshape(shape)
+
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_uint8.py:82]: The scalar is passed as `float(scalar)`. For UINT8 relational ops, the meaningful scalar range is `[0, 255]` (integers). Passing `float(0)` / `float(255)` is fine here, but worth adding a comment or assertion that the scalar should be in `[0, 255]` to prevent someone later adding e.g. `scalar=300` which would be out-of-range for uint8 semantics.

**Review by Aswinmcw — COMMENTED** [review]
Reviewed the uint8 relational ops support. The cast-before-dispatch approach (UINT8→UINT16) is clean and avoids needing to plumb UINT8 through the SFPU init/format paths. The `relational_binary` template was a thin wrapper around `invoke_binary_ng`, so the macro consolidation is a safe refactor. Test coverage looks solid with tensor-tensor and tensor-scalar paths, corner cases, and BH sim skip.

A few inline notes below — nothing blocking.

**Review by Aswinmcw — APPROVED** [review]
All prior review comments addressed:
- Shallow Tensor copy on non-UINT8 path → fixed with `std::optional<Tensor>` + reference pattern, zero-cost for existing dtypes
- Comment accuracy in `binary_ng_program_factory.cpp` → updated to distinguish EQ/NE (SUB+postprocess) from LT/GT/LE/GE (SFPU comparison)
- Single-entry parametrize noise in test → inlined as constants, scalar range comment added

LGTM.
